mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,224 @@
+
+        noisy = num_circuits * num_methods * num_options
+        ideal = 0  # TODO: ideal executor is currently unused
+        return {
+            "noisy_executions": noisy,
+            "ideal_executions": ideal,
+        }
+
+    def run_circuits(self) -> List[Dict[str, Any]]:
+        """Run the calibration circuits and store the ideal and noisy
+        expectation values for each circuit in `self.results`."""
+        expvals = []
+        for problem in self.circuits:
+            circuit, distribution = problem["circuit", "ideal_distribution"]
+            circuit.append(cirq.measure(circuit.all_qubits()))
+
+            expval_executor, bitstring_to_measure = convert_to_expval_executor(
+                self.executor, distribution
+            )
+            noisy_value = expval_executor.evaluate(circuit)[0]
+            ideal_value = distribution[bitstring_to_measure]
+            noisy_error = ideal_value - noisy_value
+
+            mitigated_values: Dict[str, Dict[str, Any]] = {
+                technique: {"results": [], "improvement_factor": None}
+                for technique in self.settings.techniques
+            }
+            for strategy in self.settings.make_strategies():
+                technique, params, mitiq_func = strategy
+                mitigated_value = mitiq_func(circuit, expval_executor)
+
+                improvement_factor = abs(
+                    noisy_error / (ideal_value - mitigated_value)
+                )
+                result = {
+                    "mitigated_value": mitigated_value,
+                    "improvement_factor": improvement_factor,
+                    "strategy": strategy,
+                    **params,
+                }
+                mitigated_values[technique]["results"].append(result)
+
+            circuit_info = asdict(problem)
+            # remove circuit; it can be regenerated if needed
+            del circuit_info["circuit"]
+            expvals.append(
+                {
+                    "circuit_info": circuit_info,
+                    "noisy_value": noisy_value,
+                    "ideal_value": distribution[bitstring_to_measure],
+                    "mitigated_values": mitigated_values,
+                }
+            )
+        return expvals
+
+    def compute_improvements(
+        self, experiment_results: List[Dict[str, Any]]
+    ) -> None:
+        """Computes the improvement factors for each calibration circuit that
+        was run. Saves the improvement factors in the input dictionary."""
+        for result in experiment_results:
+            ideal_value = result["ideal_value"]
+            noisy_value = result["noisy_value"]
+            for di in result["mitigated_values"].values():
+                results = di["results"]
+                mitigated_values = list(
+                    map(lambda di: di["mitigated_value"], results)
+                )
+                improvement_factor = abs(noisy_value - ideal_value) / sqrt(
```

Comment:
This is something important for next year. Introducing trials, i.e. repetitions of the same experiment with the same strategy.
```suggestion
                TODO: Use root-mean-square error over multiple trials for equal settings.
                improvement_factor = abs(noisy_value - ideal_value) / sqrt(
```